libcloud/httplib_ssl.py

```diff
@@ -292,21 +292,50 @@ def connect(self):
 
         ssl_version = libcloud.security.SSL_VERSION
 
+        # Connect, using contexts and SNI if we can
+
         try:
-            self.sock = ssl.wrap_socket(
-                sock,
-                self.key_file,
-                self.cert_file,
-                cert_reqs=ssl.CERT_REQUIRED,
-                ca_certs=self.ca_cert,
-                ssl_version=ssl_version)
-        except socket.error:
-            exc = sys.exc_info()[1]
-            # Re-throw an exception with a more friendly error message
-            exc = get_socket_error_exception(ssl_version=ssl_version, exc=exc)
-            raise exc
+            self.tls_context = ssl.SSLContext(ssl_version)
+            self.tls_context.verify_mode = ssl.CERT_REQUIRED
+            if self.cert_file and self.key_file:
+                self.tls_context.load_cert_chain(
+                    self.cert_file, self.key_file, None)
+            if self.ca_cert:
+                self.tls_context.load_verify_locations(self.ca_cert)
+        except AttributeError:
+            self.tls_context = None
+
+        if self.tls_context and ssl.HAS_SNI:
```

Comment:
And maybe even put this check at the top and only try to instantiate `ssl.SSLContext`, etc. if `has_sni is True`.

I know this is a style / preference thing, but I found current approach a bit confusing / hard to understand. Making it very clear what is going on (we only try to instantiate `ssl.SSLContext` and use SNI when `has_sni` is True aka this functionality is available, etc.) would make it more clear, imo.

And I think this is especially important since it's security sensitive code - we should try to make it as simple and obvious / easy to read and understand as possible :)